Gradle task fails to copy over sourceset jar properly

### Describe the bug

After upgrading to Quarkus 3.24.2 we started having failures running our application.

When running 

```sh
./gradlew quarkusRun
```

The application fails to run with the following error
```
14:47:06,458 ERROR [io.qua.run.Quarkus] Error running Quarkus: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
        at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:501)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:79)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:69)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:37)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 8 more
Caused by: java.lang.RuntimeException: java.nio.file.NoSuchFileException: /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/quarkus-app/lib/main/86495d84d99483297caec829b5c55a19e46cea95.generatedApi-1751578994559.jar
        at io.quarkus.bootstrap.runner.JarFileReference.asyncLoadAcquiredJarFile(JarFileReference.java:209)
        at io.quarkus.bootstrap.runner.JarFileReference.withJarFile(JarFileReference.java:147)
        at io.quarkus.bootstrap.runner.JarResource.getResourceData(JarResource.java:56)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:106)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:72)
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.defineClass(RunnerClassLoader.java:152)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:111)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:72)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:534)
        at java.base/java.lang.Class.forName(Class.java:513)
        at my.groupId.PetsResource_Bean.<init>(Unknown Source)
        at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Unknown Source)
        at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
        at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:129)
        at io.quarkus.arc.Arc.initialize(Arc.java:38)
        at io.quarkus.arc.runtime.ArcRecorder.initContainer(ArcRecorder.java:52)
        at io.quarkus.runner.recorded.ArcProcessor$initializeContainer1770303700.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.ArcProcessor$initializeContainer1770303700.deploy(Unknown Source)
        ... 9 more
Caused by: java.nio.file.NoSuchFileException: /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/quarkus-app/lib/main/86495d84d99483297caec829b5c55a19e46cea95.generatedApi-1751578994559.jar
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
        at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1854)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1445)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:724)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:251)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:180)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:345)
        at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:49)
        at io.quarkus.bootstrap.runner.JarFileReference.asyncLoadAcquiredJarFile(JarFileReference.java:207)
        ... 29 more
```


This bug isn't present in `3.24.1`. It `quarkusRun` passes with the following properties:
```
# Gradle properties
quarkusPluginId=io.quarkus
quarkusPluginVersion=3.24.1
quarkusPlatformGroupId=io.quarkus.platform
quarkusPlatformArtifactId=quarkus-bom
quarkusPlatformVersion=3.24.1
```

The key thing the gradle build script is doing is generating Resources and DTOs from an [openapi.yaml schema](https://github.com/TSFenwick/sourceset-quarkus-gradle-repro/blob/main/src/main/resources/openapi/openapi.yaml) into a separate sourceset [generatedApi](https://github.com/TSFenwick/sourceset-quarkus-gradle-repro/blob/main/build.gradle.kts#L34-L40) that the sourceset main depends upon.

### Expected behavior

Application runs properly with no exception

```
./gradlew quarkusRun

> Task :openApiGenerate
############################################################################################
# Thanks for using OpenAPI Generator.                                                      #
# We appreciate your support! Please consider donation to help us maintain this project.   #
# https://opencollective.com/openapi_generator/donate                                      #
############################################################################################
Successfully generated code to /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/generated-src
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2025-07-03 15:04:32,508 INFO  [io.quarkus] (main) sourceset-quarkus-gradle-repro 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.24.1) started in 0.275s. Listening on: http://0.0.0.0:8080
2025-07-03 15:04:32,510 INFO  [io.quarkus] (main) Profile prod activated. 
2025-07-03 15:04:32,510 INFO  [io.quarkus] (main) Installed features: [cdi, hibernate-validator, reactive-routes, rest, rest-jackson, smallrye-context-propagation, vertx]
```

### Actual behavior

When running 

```sh
./gradlew quarkusRun
```

The application fails to run with the following error
```
14:47:06,458 ERROR [io.qua.run.Quarkus] Error running Quarkus: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
        at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:501)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:79)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:69)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:37)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 8 more
Caused by: java.lang.RuntimeException: java.nio.file.NoSuchFileException: /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/quarkus-app/lib/main/86495d84d99483297caec829b5c55a19e46cea95.generatedApi-1751578994559.jar
        at io.quarkus.bootstrap.runner.JarFileReference.asyncLoadAcquiredJarFile(JarFileReference.java:209)
        at io.quarkus.bootstrap.runner.JarFileReference.withJarFile(JarFileReference.java:147)
        at io.quarkus.bootstrap.runner.JarResource.getResourceData(JarResource.java:56)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:106)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:72)
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.defineClass(RunnerClassLoader.java:152)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:111)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:72)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:534)
        at java.base/java.lang.Class.forName(Class.java:513)
        at my.groupId.PetsResource_Bean.<init>(Unknown Source)
        at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Unknown Source)
        at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
        at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:129)
        at io.quarkus.arc.Arc.initialize(Arc.java:38)
        at io.quarkus.arc.runtime.ArcRecorder.initContainer(ArcRecorder.java:52)
        at io.quarkus.runner.recorded.ArcProcessor$initializeContainer1770303700.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.ArcProcessor$initializeContainer1770303700.deploy(Unknown Source)
        ... 9 more
Caused by: java.nio.file.NoSuchFileException: /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/quarkus-app/lib/main/86495d84d99483297caec829b5c55a19e46cea95.generatedApi-1751578994559.jar
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
        at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1854)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1445)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:724)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:251)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:180)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:345)
        at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:49)
        at io.quarkus.bootstrap.runner.JarFileReference.asyncLoadAcquiredJarFile(JarFileReference.java:207)
        ... 29 more
```


The task `quarkusDependenciesBuild` isn't copying the correct jar name over.
```
2025-07-03T14:56:32.233-0700 [DEBUG] [org.gradle.api.Task] Dependency 86495d84d99483297caec829b5c55a19e46cea95:generatedApi::jar:1751579790328 : copying /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/classes/java/generatedApi to /Users/tomfenwick/Dev/github.com/tsfenwick/sourceset-quarkus-gradle-repro/build/quarkus-build/dep/lib/main/86495d84d99483297caec829b5c55a19e46cea95.generatedApi
```
 is looking for `86495d84d99483297caec829b5c55a19e46cea95.generatedApi-1751578994559.jar` to run successfully

[build scan](https://scans.gradle.com/s/ntjldmjtsoqcw)

`./gradlew quarkusRun` with info and debug logging

[gradle-info.txt](https://github.com/user-attachments/files/21047473/gradle-info.txt)
[gradle-debug.txt](https://github.com/user-attachments/files/21047472/gradle-debug.txt)




### How to Reproduce?

https://github.com/TSFenwick/sourceset-quarkus-gradle-repro

run `./gradlew quarkusRun`

### Output of `uname -a` or `ver`

Darwin Tom-Fenwick 24.5.0 Darwin Kernel Version 24.5.0: Tue Apr 22 19:53:27 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T6041 arm64

### Output of `java -version`

openjdk 21.0.7 2025-04-15 LTS OpenJDK Runtime Environment Zulu21.42+19-CA (build 21.0.7+6-LTS) OpenJDK 64-Bit Server VM Zulu21.42+19-CA (build 21.0.7+6-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.24.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.14.2

### Additional information

Scanning through some of the changes for 3.24.2. This failure could be related to https://github.com/quarkusio/quarkus/pull/48551

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
@@ -920,10 +920,7 @@ public class JarResultBuildStep {
             return;
         }
         for (Path resolvedDep : appDep.getResolvedPaths()) {
-            final boolean isDirectory = Files.isDirectory(resolvedDep);
-            // we don't use getFileName() for directories, since directories would often be "classes" ending up merging content from multiple dependencies in the same package
-            final String fileName = isDirectory ? getFileNameForDirectory(appDep)
-                    : appDep.getGroupId() + "." + resolvedDep.getFileName();
+            final String fileName = getJarFileName(appDep, resolvedDep);
             final Path targetPath;
 
             if (allowParentFirst && parentFirstArtifacts.contains(appDep.getKey())) {
@@ -935,7 +932,7 @@ public class JarResultBuildStep {
             }
             runtimeArtifacts.computeIfAbsent(appDep.getKey(), (s) -> new ArrayList<>(1)).add(targetPath);
 
-            if (isDirectory) {
+            if (Files.isDirectory(resolvedDep)) {
                 // This case can happen when we are building a jar from inside the Quarkus repository
                 // and Quarkus Bootstrap's localProjectDiscovery has been set to true. In such a case
                 // the non-jar dependencies are the Quarkus dependencies picked up on the file system
@@ -976,12 +973,20 @@ public class JarResultBuildStep {
     }
 
     /**
-     * Returns a JAR file name to be used for a content of a dependency that is in a directory.
+     * Returns a JAR file name to be used for a content of a dependency, depending on whether the resolved path
+     * is a directory or not.
+     * We don't use getFileName() for directories, since directories would often be "classes" ending up merging
+     * content from multiple dependencies in the same package
      *
      * @param dep dependency
+     * @param resolvedPath path of the resolved dependency
      * @return JAR file name
      */
-    private static String getFileNameForDirectory(ResolvedDependency dep) {
+    public static String getJarFileName(ResolvedDependency dep, Path resolvedPath) {
+        boolean isDirectory = Files.isDirectory(resolvedPath);
+        if (!isDirectory) {
+            return dep.getGroupId() + "." + resolvedPath.getFileName();
+        }
         final StringBuilder sb = new StringBuilder();
         sb.append(dep.getGroupId()).append(".").append(dep.getArtifactId()).append("-");
         if (!dep.getClassifier().isEmpty()) {

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusBuildDependencies.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusBuildDependencies.java
@@ -19,6 +19,7 @@ import org.gradle.api.tasks.TaskAction;
 
 import io.quarkus.bootstrap.model.ApplicationModel;
 import io.quarkus.deployment.pkg.PackageConfig;
+import io.quarkus.deployment.pkg.steps.JarResultBuildStep;
 import io.quarkus.fs.util.ZipUtils;
 import io.quarkus.gradle.QuarkusPlugin;
 import io.quarkus.maven.dependency.ArtifactKey;
@@ -184,7 +185,7 @@ public abstract class QuarkusBuildDependencies extends QuarkusBuildTask {
                     ResolvedDependency dep = depAndTarget.getValue();
                     Path targetDir = depAndTarget.getKey();
                     dep.getResolvedPaths().forEach(p -> {
-                        String file = dep.getGroupId() + '.' + p.getFileName();
+                        String file = JarResultBuildStep.getJarFileName(dep, p);
                         Path target = targetDir.resolve(file);
                         if (!Files.exists(target)) {
                             getLogger().debug("Dependency {} : copying {} to {}",

--- a/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/build.gradle
+++ b/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/build.gradle
@@ -1,0 +1,33 @@
+plugins {
+    id 'java'
+    id 'io.quarkus'
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex 'io.quarkus.*'
+            includeGroup 'org.hibernate.orm'
+        }
+    }
+    mavenCentral()
+    gradlePluginPortal()
+}
+
+sourceSets {
+    additional {
+        java {
+            srcDir "src/gen/java"
+        }
+    }
+}
+
+dependencies {
+    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
+    implementation sourceSets.additional.output
+    implementation 'io.quarkus:quarkus-arc'
+    implementation 'io.quarkus:quarkus-rest'
+}
+
+group 'org.acme'
+version '1.0.0-SNAPSHOT'

--- a/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/gradle.properties
@@ -1,0 +1,2 @@
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus

--- a/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/settings.gradle
+++ b/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/settings.gradle
@@ -1,0 +1,16 @@
+pluginManagement {
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex 'io.quarkus.*'
+                includeGroup 'org.hibernate.orm'
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+      id 'io.quarkus' version "${quarkusPluginVersion}"
+    }
+}
+rootProject.name = 'additional-source-set-as-dependency'

--- a/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/src/gen/java/io/reproducer/GreetApi.java
+++ b/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/src/gen/java/io/reproducer/GreetApi.java
@@ -1,0 +1,6 @@
+package io.reproducer;
+
+public interface GreetApi {
+
+  public String hello();
+}

--- a/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/src/main/java/org/acme/GreetingResource.java
+++ b/integration-tests/gradle/src/main/resources/additional-source-set-as-dependency/src/main/java/org/acme/GreetingResource.java
@@ -1,0 +1,17 @@
+package org.acme;
+
+import io.reproducer.GreetApi;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+@Path("/hello")
+public class GreetingResource implements GreetApi {
+
+  @GET
+  @Produces(MediaType.TEXT_PLAIN)
+  public String hello() {
+    return "Hello from Quarkus REST...";
+  }
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/run/AdditionalSourceSetAsDependencyTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/run/AdditionalSourceSetAsDependencyTest.java
@@ -1,0 +1,28 @@
+package io.quarkus.gradle.run;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.quarkus.gradle.devmode.QuarkusDevGradleTestBase;
+
+// Reproduces https://github.com/quarkusio/quarkus/issues/48768.
+// Despite extending QuarkusDevGradleTestBase, this test does not use the dev mode,
+// but rather runs the application in a production mode using `quarkusRun` task.
+// At this point it's the only test that does so, but if more such tests are added,
+// perhaps it would make sense to create a separate base class for them.
+public class AdditionalSourceSetAsDependencyTest extends QuarkusDevGradleTestBase {
+
+    @Override
+    protected String projectDirectoryName() {
+        return "additional-source-set-as-dependency";
+    }
+
+    @Override
+    protected String[] buildArguments() {
+        return new String[] { "clean", "quarkusRun" };
+    }
+
+    @Override
+    protected void testDevMode() throws Exception {
+        assertThat(getHttpResponse("/hello")).contains("Hello from Quarkus REST...");
+    }
+}

